### PR TITLE
Forbid uncalled functions in evaluated output

### LINKED ISSUE #158: `$json` should handle functions with an error -`{$json: {$eval: 'min'}}`
`{$json: {$eval: 'min'}}` is the test case, it produces some weird error that isn't handled.
I think we want to explicitly check for this...

### PR DESCRIPTION
**Pull request by mkantor**
Emit an error on uncalled functions in all implementations.

Fixes #158.

# Checklist

Before submitting a pull request, please check the following:

* [x] All tests pass for you on your machine for all implementations (all implementations must behave identically)
* [x] New tests have been added for any new functionality or to prevent regressions of a bugfix
* [x] Added a short description of the change to `newsfragments/$issue.bugfix` (for fixes) or `.feature` (for new features) or `.doc` (for documentation-only changes)

### CODE REVIEW

**Comment by mkantor** [comment]
Unfortunately because functions can be passed around as regular values during evaluation (such as in [the "dotted function expression" test case](https://github.com/taskcluster/json-e/blob/63c6a7f6d42eced73f5cf0294e873faeb76acb3d/specification.yml#L4255-L4258)) the interpreters can't simply abort when they encounter an uncalled function. The best I was able to come up with was to recursively validate the final evaluated object, but if anyone has other ideas I'm all ears.

File: js/src/index.js [hunk 1]
```diff
@@ -474,6 +478,25 @@ let parseUntilTerminator = (source, terminator, context) => {
...
 };
 
+let containsFunctions = (rendered) => {
+  if (isFunction(rendered)) {
+    return true;
+  } else if (Array.isArray(rendered)) {
+    return rendered.some(containsFunctions);
+  } else if (typeof rendered === 'object' && rendered !== null) {
+    for (const key in rendered) {
+      if (rendered.hasOwnProperty(key)) {
```
Comment (djmitche) [inline js/src/index.js:488]: another option for this kind of loop is `for (const key of Object.keys(rendered))`

**Review by djmitche — APPROVED** [review]
This looks good!  If you would like to make the change below, let me know, but I'm also happy to merge it as-is too.

File: js/src/index.js (hunk 1 above)
Comment (mkantor) [reply js/src/index.js]: Thanks, I made the suggested edit and force-pushed this branch.

**Review by djmitche — APPROVED** [review]

**Comment by djmitche** [comment]
Fantastic, thank you for that!

**Pull request merged**